### PR TITLE
Bobshie beforeconfig skip

### LINKED ISSUE #2880: before configuration and before invocation should be 'SKIP' when beforeMethod is 'skip'
7.6.1

### Expected behavior

```
16:02:12.583 [main] INFO  TestNgDemoListener - after invocation executed, method:beforeClass,testResultStatus:FAILURE/2
java.lang.ArithmeticException: / by zero
16:02:12.631 [main] INFO  TestNgDemoListener - before configuration executed, method:beforeMethod,testResultStatus:**SKIP/3**
16:02:12.631 [main] INFO  TestNgDemoListener - before invocation executed, method:beforeMethod,testResultStatus:**SKIP/3**
16:02:12.632 [main] INFO  TestNgDemoListener - after invocation executed, method:beforeMethod,testResultStatus:SKIP/3
16:02:12.633 [main] INFO  TestNgDemoListener - Skip Config Method,beforeMethod,testResultStatus:SKIP/3
14:03:16.659 [main] INFO  TestNgDemoListener - onTestStart executed, method:issueTest,testResultStatus:STARTED/16
14:03:16.661 [main] INFO  TestNgDemoListener - onTestSkipped executed, method:issueTest,testResultStatus:SKIP/3
14:03:16.660 [main] INFO  TestNgDemoListener - before invocation executed, method:issueTest,testResultStatus:SKIP/3
14:03:16.660 [main] INFO  TestNgDemoListener - after invocation executed, method:issueTest,testResultStatus:SKIP/3
```

### Actual behavior

```
16:02:12.583 [main] INFO  TestNgDemoListener - after invocation executed, method:beforeClass,testResultStatus:FAILURE/2
java.lang.ArithmeticException: / by zero
16:02:12.631 [main] INFO  TestNgDemoListener - before configuration executed, method:beforeMethod,testResultStatus:**STARTED/16**
16:02:12.631 [main] INFO  TestNgDemoListener - before invocation executed, method:beforeMethod,testResultStatus:**STARTED/16**
16:02:12.632 [main] INFO  TestNgDemoListener - after invocation executed, method:beforeMethod,testResultStatus:SKIP/3
16:02:12.633 [main] INFO  TestNgDemoListener - Skip Config Method,beforeMethod,testResultStatus:SKIP/3
14:03:16.659 [main] INFO  TestNgDemoListener - onTestStart executed, method:issueTest,testResultStatus:STARTED/16
14:03:16.661 [main] INFO  TestNgDemoListener - onTestSkipped executed, method:issueTest,testResultStatus:SKIP/3
14:03:16.660 [main] INFO  TestNgDemoListener - before invocation executed, method:issueTest,testResultStatus:**SKIP/3**
14:03:16.660 [main] INFO  TestNgDemoListener - after invocation executed, method:issueTest,testResultStatus:SKIP/3
```

### Is the issue reproducible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans

Simply use listener:

```java
public class TestNgDemoListener implements IInvokedMethodListener, IConfigurationListener, ITestListener

import org.slf4j.LoggerFactory;
import org.testng.IConfigurationListener;
import org.testng.IInvokedMethod;
import org.testng.IInvokedMethodListener;
import org.testng.ITestListener;
import org.testng.ITestNGMethod;
import org.testng.ITestResult;

public class TestNgDemoListener implements IInvokedMethodListener, IConfigurationListener, ITestListener {
    org.slf4j.Logger logger = LoggerFactory.getLogger(TestNgDemoListener.class);

    @Override
    public void beforeInvocation(IInvokedMethod method, ITestResult testResult) {
        logger.info(
                "before invocation executed, method:" + testResult.getMethod().getMethodName() + ",testResultStatus:"
                        + getStatusString(testResult.getStatus()));

    }

    @Override
    public void beforeConfiguration(ITestResult testResult) {
        logger.info(
                "before configuration executed, method:" + testResult.getMethod().getMethodName() + ",testResultStatus:"
                        + getStatusString(testResult.getStatus()));
    }

    @Override
    public void onConfigurationSkip(ITestResult testResult, ITestNGMethod method) {
        logger.info("Skip Config Method," + testResult.getMethod().getMethodName() + ",testResultStatus:"
                + getStatusString(testResult.getStatus()));
    }

    @Override
    public void afterInvocation(IInvokedMethod method, ITestResult testResult) {
        logger.info(
                "after invocation executed, method:" + testResult.getMethod().getMethodName() + ",testResultStatus:"
                        + getStatusString(testResult.getStatus()));
    }

    @Override
    public void onTestStart(ITestResult result) {
        logger.info(
                "onTestStart executed, method:" + result.getMethod().getMethodName() + ",testResultStatus:"
                        + getStatusString(result.getStatus()));
    }

    @Override
    public void onTestSkipped(ITestResult result) {
        logger.info(
                "onTestSkipped executed, method:" + result.getMethod().getMethodName() + ",testResultStatus:"
                        + getStatusString(result.getStatus()));
    }

    public static String getStatusString(int status) {
        return switch (status) {
        case -1 -> "CREATED/" + status;
        case 1 -> "SUCCESS/" + status;
        case 2 -> "FAILURE/" + status;
        case 3 -> "SKIP/" + status;
        case 4 -> "SUCCESS_PERCENTAGE_FAILURE/" + status;
        case 16 -> "STARTED/" + status;
        default -> "Others/" + status;
        };
    }
}
```

```java
import org.slf4j.LoggerFactory;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeGroups;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Listeners;
import org.testng.annotations.Test;

@Listeners({ TestNgDemoListener.class})
public class TestNgDemoTest {
    org.slf4j.Logger logger = LoggerFactory.getLogger(TestNgDemoTest.class);

    @BeforeClass
    public void beforeClass() {
        logger.info("beforeClass");
        int i = 5/0; // throw an exception
    }

    @BeforeMethod
    public void beforeMethod() {
        logger.info("beforeMethod");
    }

    @BeforeTest
    public void beforeTest() {
        logger.info("beforeTest");
    }

    @BeforeSuite
    public void beforeSuite() {
        logger.info("beforeSuite");
    }

    @BeforeGroups
    public void beforeGroups(){
        logger.info("beforeGroups");
    }


    @Test
    public void issueTest() {
        logger.info("Test BeforeConfiguration behavior");
    }

}
```

### Contribution guidelines

Incase you plan to raise a pull request to fix this issue, please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

### Description
After one of our test project uplift TestNG from 7.5 to 7.6 version, there are some unexpected behavior from before. the related issue is: 
[beforeConfiguration() listener method should be invoked for skipped configurations as well · Issue #2729 · cbeust/testng · GitHub](https://github.com/cbeust/testng/issues/2729#issuecomment-1041209003)

### PR DESCRIPTION
**Pull request by bobshie**
Fixes #2880 .

### Did you remember to?

- [*] Add test case(s)
- [*] Update `CHANGES.txt`
- [*] Auto applied styling via `./gradlew autostyleApply`

We encourage pull requests that:

* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

**Note:** For more information on contribution guidelines  please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

### CODE REVIEW

File: testng-core/src/test/java/test/listeners/issue2880/IssueTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+import org.testng.TestNG;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlSuite;
+import test.SimpleBaseTest;
+
+public class IssueTest extends SimpleBaseTest {
```
Comment (krmahadevan) [inline testng-core/src/test/java/test/listeners/issue2880/IssueTest.java:13]: Please move this test method into `test.listeners.ListenersTest`. we are trying to ensure that all the tests reside in the same class. The test classes that this test uses can reside in its own package (`issue2880` in this case)

File: testng-core/src/test/java/test/listeners/github1602/IssueTest.java [hunk 2]
```diff
@@ -34,17 +34,22 @@ public Object[][] getData() {
...
             "BeforeInvocation_afterMethod_STARTED",
             "AfterInvocation_afterMethod_SUCCESS");
+
     List<String> baseList =
         Arrays.asList(
             "BeforeInvocation_beforeMethod_STARTED",
             "AfterInvocation_beforeMethod_FAILURE",
             "BeforeInvocation_testMethod_SKIP",
-            "AfterInvocation_testMethod_SKIP",
-            "BeforeInvocation_afterMethod_STARTED");
```
Comment (krmahadevan) [inline testng-core/src/test/java/test/listeners/github1602/IssueTest.java:43]: Why did we have to remove this line ?

**Review by krmahadevan — CHANGES_REQUESTED** [review]
Have shared some comments

**Review by juherr — COMMENTED** [review]
Lgtm once approved by @krmahadevan

**Comment by Hkh9966** [comment]
image
One UT failed, but it doesn't seem to be related to this change.

File: testng-core/src/test/java/test/listeners/github1602/IssueTest.java (hunk 2 above)
Comment (bobshie) [reply testng-core/src/test/java/test/listeners/github1602/IssueTest.java:43]: hello, I changed "BeforeInvocation_afterMethod_STARTED"); to skipList.add("BeforeInvocation_afterMethod_SKIP");

in skiplist and failList there is different. 

    List<String> skipList = Lists.newArrayList(baseList);
    skipList.add("BeforeInvocation_afterMethod_SKIP");
    skipList.add("AfterInvocation_afterMethod_SKIP");

    List<String> failList = Lists.newArrayList(baseList);
    failList.add("BeforeInvocation_afterMethod_STARTED");
    failList.add("AfterInvocation_afterMethod_FAILURE");

File: testng-core/src/test/java/test/listeners/issue2880/IssueTest.java (hunk 1 above)
Comment (bobshie) [reply testng-core/src/test/java/test/listeners/issue2880/IssueTest.java]: Accepted and Done.

**Review by krmahadevan — APPROVED** [review]

**Pull request merged**